aten/src/ATen/native/cuda/IndexKernel.cu

```diff
@@ -314,50 +269,145 @@ Tensor & masked_select_out_cuda(Tensor & result, const Tensor & self, const Tens
   return masked_select_out_cuda_impl(result, self, mask);
 }
 
-void take_out_cuda_template(Tensor& output, const Tensor& input, const Tensor& index) {
-  TORCH_CHECK(output.device().type() == at::kCUDA, "device type of output (", output.device().type(), ") is not GPU");
-  TORCH_CHECK(input.device().type() == at::kCUDA, "device type of input (", input.device().type(), ") is not GPU");
-  TORCH_CHECK(index.device().type() == at::kCUDA, "device type of index (", index.device().type(), ") is not GPU");
-
-  TORCH_CHECK(output.layout() == Layout::Strided, "take() only supports strided layout, got layout: ", output.layout(), " on output tensor");
-  TORCH_CHECK(input.layout() == Layout::Strided, "take() only supports strided layout, got layout: ", input.layout(), " on input tensor");
-  TORCH_CHECK(index.layout() == Layout::Strided, "take() only supports strided layout, got layout: ", index.layout(), " on index tensor");
+template <typename scalar_t, typename index_t, int S, typename func_t>
+void cuda_take_put_kernel(
+  TensorIterator& iter,
+  const Tensor& indexed,
+  const func_t& f) {
+  if (!iter.can_use_32bit_indexing()) {
+    for (auto& sub_iter : iter.with_32bit_indexing()) {
+      cuda_take_put_kernel<scalar_t, index_t, S>(sub_iter, indexed, f);
+    }
+    return;
+  }
 
-  TORCH_CHECK(output.scalar_type() == input.scalar_type(),
-          "output and input scalar type must match. but got different types: ", output.scalar_type(), " and ", input.scalar_type());
-  TORCH_CHECK(index.scalar_type() == kLong, "index must be an int64 tensor");
+  const auto numel = indexed.numel();
+  const bool is_contiguous = indexed.is_contiguous();
 
-  TensorArg output_arg{ output, "output", 1 };
-  TensorArg input_arg{ input, "input", 2 };
-  TensorArg index_arg{ index, "index", 3 };
-  checkAllSameGPU("take", {output_arg, input_arg, index_arg});
+  char* __restrict__ iterated_ptr = reinterpret_cast<char*>(iter.data_ptr(0));
+  char* __restrict__ idx_ptr = reinterpret_cast<char*>(iter.data_ptr(1));
 
-  TORCH_CHECK(input.dim() < MAX_CUTORCH_DIMS, CUTORCH_DIM_WARNING);
-  TORCH_CHECK(output.dim() < MAX_CUTORCH_DIMS, CUTORCH_DIM_WARNING);
-  TORCH_CHECK(index.dim() < MAX_CUTORCH_DIMS, CUTORCH_DIM_WARNING);
+  const auto offset_calc = make_offset_calculator<2>(iter);
+  using uindex_t = std::make_unsigned_t<index_t>;
 
-  TORCH_CHECK(!(input.numel() == 0 && index.numel() != 0), "tried to take from an empty tensor");
+  const auto indexed_dim = indexed.dim();
+  auto indexed_sizes = indexed.sizes().vec();
```

Comment:
nit: you can create vector and flip in one go:
auto indexed_sizes = std::vector<int64_t>(indexed.sizes().rbegin(), indexed.sizes().rend());

Reply:
Is it efficient (or just fine) to call `indexed.sizes()` twice?

Reply:
Leave it to compiler to optimize away, also, indexed.sizes() returns a reference to a SmallVec that holds sizes, so it's fast